core/src/main/java/org/renjin/stats/internals/distributions/RNG.java

```diff
@@ -87,202 +89,342 @@ public static void set_seed(@Current Context context, int seed, SEXP kind, SEXP
+      if (j == shape.length()-1) j = 0;
+      if (k == scale.length()-1) k = 0;
+      vb.add(Weibull.rweibull(context.getSession(), shape.getElementAsDouble(j), scale.getElementAsDouble(k)));
+      if (shape.length() > 1) j++;
+      if (scale.length() > 1) k++;
     }
     return (vb.build());
   }
 
   @Internal
-  public static DoubleVector rnbinom(@Current Context context, int n, double size, double prob) {
-    DoubleArrayVector.Builder vb = new DoubleArrayVector.Builder();
+  public static DoubleVector rnbinom(@Current Context context, Vector nVector, DoubleVector size, DoubleVector prob) {
+    int n = defineSize(nVector);
+    DoubleArrayVector.Builder vb = DoubleArrayVector.Builder.withInitialCapacity(n);
+    if (size.length() == 0 || prob.length() == 0) return (DoubleArrayVector.Builder.withInitialSize(n).build());
+    int j = 0, k = 0;
     for (int i = 0; i < n; i++) {
-      vb.add(NegativeBinom.rnbinom(context.getSession(), size, prob));
+      if (j == size.length()-1) j = 0;
+      if (k == prob.length()-1) k = 0;
+      vb.add(NegativeBinom.rnbinom(context.getSession(), size.getElementAsDouble(j), prob.getElementAsDouble(k)));
+      if (size.length() > 1) j++;
+      if (prob.length() > 1) k++;
     }
     return (vb.build());
   }
 
   @Internal
-  public static DoubleVector rnbinom_mu(@Current Context context, int n, double size, double mu) {
-    DoubleArrayVector.Builder vb = new DoubleArrayVector.Builder();
+  public static DoubleVector rnbinom_mu(@Current Context context, Vector nVector, DoubleVector size, DoubleVector mu) {
+    int n = defineSize(nVector);
+    DoubleArrayVector.Builder vb = DoubleArrayVector.Builder.withInitialCapacity(n);
+    if (size.length() == 0 || mu.length() == 0) return (DoubleArrayVector.Builder.withInitialSize(n).build());
+    int j = 0, k = 0;
     for (int i = 0; i < n; i++) {
-      vb.add(NegativeBinom.rnbinom_mu(context.getSession(), size, mu));
+      if (j == size.length()-1) j = 0;
+      if (k == mu.length()-1) k = 0;
+      vb.add(NegativeBinom.rnbinom_mu(context.getSession(), size.getElementAsDouble(j), mu.getElementAsDouble(k)));
+      if (size.length() > 1) j++;
+      if (mu.length() > 1) k++;
     }
     return (vb.build());
   }
 
   @Internal
-  public static DoubleVector rbinom(@Current Context context, int n, double size, double prob) {
-    DoubleArrayVector.Builder vb = new DoubleArrayVector.Builder();
+  public static DoubleVector rbinom(@Current Context context, Vector nVector, DoubleVector size, DoubleVector prob) {
+    int n = defineSize(nVector);
+    DoubleArrayVector.Builder vb = DoubleArrayVector.Builder.withInitialCapacity(n);
+    if (size.length() == 0 || prob.length() == 0) return (DoubleArrayVector.Builder.withInitialSize(n).build());
+    int j = 0, k = 0;
     for (int i = 0; i < n; i++) {
-      vb.add(Binom.rbinom(context.getSession(), size, prob));
+      if (j == size.length()-1) j = 0;
```

Comment:
This doesn't seem correct. If `size.length() == 4` and `j == 3`, then you reset the counter and skip the last element. You should do this check after incrementing, not before.